### PR TITLE
Implement Secure Engine TRNG hwrandom support, add PINCTRL/GPIO support

### PR DESCRIPTION
**Pull request by alexhorner**
- Implements crypto and hwrng support for the Secure Engine
- Maps all Secure Engine register offsets and bitmaps for future expansion of Secure Engine support
- This PR also contains the previous PINCTRL/GPIO work:
- - Pullup/down
- - Matched U-Boot device tree as close as possible for now, needs review
- - Tested, works for out and in, IRQs working too
- - **Big Warning:** If an input GPIO gets unmasked, and both Linux and M0/LP are not handling and clearing interrupts, we will get stuck in a loop and essentially lock D0 up. This should not be possible unless you poke a GPIO's IRQ to unmasked with `devmem` or badly behaving code on M0/LP doesn't remask or doesn't handle interrupts for that pin

### CODE REVIEW

**Comment by alexhorner** [comment]
Marked as draft to finish implementation of GPIO pull up / pull down / schmitt / bug with IRQs

**Comment by alexhorner** [comment]
I have now finished these final pieces, and I think we are now properly ready for a review!

File: drivers/pinctrl/pinctrl-bflb.c [hunk 1]
```diff
@@ -254,11 +256,129 @@ static int bflb_gpio_direction_output(struct gpio_chip *chip, unsigned int offse
...
+
+//AH: Configure pin electrical characteristics
+static int bflb_gpio_set_config(struct gpio_chip *chip, unsigned int offset, unsigned long config)
+{
+	struct bflb_gpio_pinctrl *pctl = gpiochip_get_data(chip);
+	enum pin_config_param param = pinconf_to_config_param(config);
+	unsigned int arg = pinconf_to_config_argument(config);
+
+	switch (param)
+	{
```
Comment (arm000) [inline drivers/pinctrl/pinctrl-bflb.c]: There's a bunch of Linux style violations here such as curly braces on their own line, can you run checkpatch to find and fix these?

**Review by arm000 — CHANGES_REQUESTED** [review]
This all looks pretty good, but it's a little hard to review because of the many fixup patches, can you squash these together into a single patch, or maybe 3: device-tree changes, add driver, update bl808_defconfig to enable?

**Comment by alexhorner** [comment]
Hey @arm000 

I have rewritten the branch to move the commits into three distinct commits. I didn't exactly squash in as much as build whole new commits with the style guidelines checks upon them.

All errors from checkpath have been fixed, but I have decided to ignore the following warnings (please let me know if I should do something about them or leave them for now):

---

```
./scripts/checkpatch.pl -f drivers/pinctrl/pinctrl-bflb.c

WARNING: Block comments use * on subsequent lines
#605: FILE: drivers/pinctrl/pinctrl-bflb.c:605:
+		FIELD_PREP(REG_GPIOx_FUNC_SEL, 11/*SWGPIO*/) //Value - Set func
+	);

WARNING: Avoid line continuations in quoted strings
#774: FILE: drivers/pinctrl/pinctrl-bflb.c:774:
+	dev_info(&pdev->dev, "Bouffalo Lab pinctrl+GPIO(+interrupt) controller - \

WARNING: DT compatible string "bflb,pinctrl" appears un-documented -- check ./Documentation/devicetree/bindings/
#783: FILE: drivers/pinctrl/pinctrl-bflb.c:783:
+	{ .compatible = "bflb,pinctrl", },

WARNING: DT compatible string vendor "bflb" appears un-documented -- check ./Documentation/devicetree/bindings/vendor-prefixes.yaml
#783: FILE: drivers/pinctrl/pinctrl-bflb.c:783:
+	{ .compatible = "bflb,pinctrl", },

total: 0 errors, 4 warnings, 802 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

drivers/pinctrl/pinctrl-bflb.c has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.
```

---

```
./scripts/checkpatch.pl -f drivers/crypto/bflb-seceng.c

WARNING: else is not generally useful after a break or return
#610: FILE: drivers/crypto/bflb-seceng.c:610:
+			break;
+		} else {

WARNING: else is not generally useful after a break or return
#679: FILE: drivers/crypto/bflb-seceng.c:679:
+			break;
+		} else {

WARNING: DT compatible string "bflb,seceng" appears un-documented -- check ./Documentation/devicetree/bindings/
#822: FILE: drivers/crypto/bflb-seceng.c:822:
+	{ .compatible = "bflb,seceng", .data = (const void *)0, },

WARNING: DT compatible string vendor "bflb" appears un-documented -- check ./Documentation/devicetree/bindings/vendor-prefixes.yaml
#822: FILE: drivers/crypto/bflb-seceng.c:822:
+	{ .compatible = "bflb,seceng", .data = (const void *)0, },

total: 0 errors, 4 warnings, 840 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

drivers/crypto/bflb-seceng.c has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.
```

Thanks!

**Comment by grant-olson** [comment]
You can use literal concatenation to break up the multiline string:

```
"This is a big long string"
```

```
"This is a" " big long string"
```

```
"This is a"
"big long string"
```

**Comment by alexhorner** [comment]
> You can use literal concatenation to break up the multiline string:
> 
> ```
> "This is a big long string"
> ```
> 
> ```
> "This is a" " big long string"
> ```
> 
> ```
> "This is a"
> "big long string"
> ```

This is certainly cleaner, so I'll roll it in with any other feedback I get, although it also produces a similar warning

```
WARNING: quoted string split across lines
#775: FILE: drivers/pinctrl/pinctrl-bflb.c:775:
+	dev_info(&pdev->dev, "Bouffalo Lab pinctrl+GPIO(+interrupt) controller - "
+			"Registered %lu function(s) for %u pin(s)",
```

File: drivers/pinctrl/pinctrl-bflb.c [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+	struct device *dev;
+	struct pinctrl_dev *pctldev;
+
+	void __iomem *base;
+	struct regmap *map;
+
+	struct pinctrl_desc pinctrl_desc;
+	struct gpio_chip gpio_chip;
+
+	bool irqsunmasked[64];
```
Comment (arm000) [inline drivers/pinctrl/pinctrl-bflb.c]: This size needs to match "bflb,npins" from device-tree and should be dynamically allocated to match

File: drivers/pinctrl/pinctrl-bflb.c [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+static int bflb_gpio_set_config(struct gpio_chip *chip, unsigned int offset,
+		unsigned long config)
+{
+	struct bflb_gpio_pinctrl *pctl = gpiochip_get_data(chip);
+	enum pin_config_param param = pinconf_to_config_param(config);
+	unsigned int arg = pinconf_to_config_argument(config);
+
+	switch (param) {
+	case PIN_CONFIG_BIAS_DISABLE:
+		bflb_gpio_set_reg
```
Comment (arm000) [inline drivers/pinctrl/pinctrl-bflb.c]: nit: LKML style encourages functions to fit on one screen, so arguments on separate lines is discouraged.

File: drivers/pinctrl/pinctrl-bflb.c [hunk 4]
```diff
@@ -0,0 +1,802 @@
...
+	case IRQ_TYPE_LEVEL_HIGH: return BFLB_IRQ_MODE_SYNC_LEVEL_HIGH;
+
+	case IRQ_TYPE_LEVEL_LOW: return BFLB_IRQ_MODE_SYNC_LEVEL_LOW;
+
+	//No "off" available on BL808, set to default IRQ_TYPE_EDGE_FALLING and
+	//then we'll need to mask
+	default: return BFLB_IRQ_MODE_SYNC_EDGE_FALLING;
+	}
+
+	return BFLB_IRQ_MODE_SYNC_EDGE_FALLING;
```
Comment (arm000) [inline drivers/pinctrl/pinctrl-bflb.c]: nit: use a local variable to capture the return type and you can avoid having multiple points of exit from this function, and this unreachable code here

File: drivers/pinctrl/pinctrl-bflb.c [hunk 5]
```diff
@@ -0,0 +1,802 @@
...
+			if (res < 0)
+				return dev_err_probe(pctl->dev, res,
+						"Failed to register function.");
+		}
+	}
+
+	dev_info(&pdev->dev, "Bouffalo Lab pinctrl+GPIO(+interrupt) controller - \
+Registered %lu function(s) for %u pin(s)",
+
+			ARRAY_SIZE(pinmux_functions), npins);
```
Comment (arm000) [inline drivers/pinctrl/pinctrl-bflb.c]: don't break strings with backslash.  Instead use two strings which the compiler will combine into one like this:

```
dev_info(&pdev->dev, "Bouffalo Lab pinctrl+GPIO(+interrupt) controller - "
                     "Registered %lu function(s) for %u pin(s)", 
                     ARRAY_SIZE(pinmux_functions), npins);
```

File: drivers/crypto/bflb-seceng.c [hunk 6]
```diff
@@ -0,0 +1,840 @@
...
+#define REG_SECENG_SHA_0_CTRL_INT                   BIT(8)
+#define REG_SECENG_SHA_0_CTRL_HASH_SEL              BIT(6)
+#define REG_SECENG_SHA_0_CTRL_EN                    BIT(5)
+#define REG_SECENG_SHA_0_CTRL_MODE                  GENMASK(4, 2)
+#define REG_SECENG_SHA_0_CTRL_INT_TRIG_1T           BIT(1)
+#define REG_SECENG_SHA_0_CTRL_BUSY                  BIT(0)
+
+//se_sha_0_msa
+#define REG_SECENG_SHA_0_MSA                        4
+#define REG_SECENG_SHA_0_MSA_VAL                    GENMASK(31, 0)
```
Comment (arm000) [inline drivers/crypto/bflb-seceng.c]: Probably don't need to use GENMASK() when the register contains no bit fields.

File: drivers/crypto/bflb-seceng.c [hunk 7]
```diff
@@ -0,0 +1,840 @@
...
+
+struct bflb_seceng {
+	struct mutex lock;
+	struct device *dev;
+	void __iomem *base;
+	struct regmap *map;
+	unsigned int initialised;
+
+	struct hwrng hwrng;
+	char hwrng_name[64];
```
Comment (arm000) [inline drivers/crypto/bflb-seceng.c]: unused?

File: drivers/crypto/bflb-seceng.c [hunk 8]
```diff
@@ -0,0 +1,840 @@
...
+	dev_dbg(bflb_seceng_dev->dev, "Refreshed TRNG");
+}
+
+static unsigned int bflb_seceng_read_rng_dout(struct bflb_seceng *seceng, unsigned int doutreg)
+{
+	unsigned int selectedreg = 0;
+
+	switch (doutreg) {
+	case 0:
+		selectedreg = REG_SECENG_TRNG_0_DOUT_0;
```
Comment (arm000) [inline drivers/crypto/bflb-seceng.c]: These registers are all sequential, so this switch/lookup is unecessary, just replace with a direct readl_relaxed() and compute the offset like this:

readl_relaxed(seceng->base + REG_SECENG_TRNG_0_DOUT_0 + offset*4);

File: drivers/crypto/bflb-seceng.c [hunk 9]
```diff
@@ -0,0 +1,840 @@
...
+	unsigned int written = 0;
+	unsigned int doutreg = 8;
+	u32 val;
+
+	mutex_lock(&bflb_seceng_dev->lock);
+
+	dev_dbg(seceng->dev, "Starting TRNG crypto read for %u bytes...", dlen);
+
+	do {
+		if (doutreg >= 8) {
```
Comment (arm000) [inline drivers/crypto/bflb-seceng.c]: This complicated loop/math can be simplified if you refactor into the following functions:
*crypto_read_multiple() - this function: calls crypto_read8() in a loop to fill in the buffer one byte at a time
*crypto_read8() - calls *crypto_read32() to get 32-bits of randomness and returns 8 bits of it.  remembers previous cached 32-bit value and returns next 8 bits of it until empty and then refreshes with another call to *crypto_read32()
*crypto_read32() - returns the next 32-bit dout value, and calls refresh() when there is no more randomness

**Review by arm000 — CHANGES_REQUESTED** [review]

**Comment by alexhorner** [comment]
Hey @arm000 

I have resolved all of the review you've given me. Please let me know if any resolutions are not satisfactory. Checkpatch has been run too.

The crypto half of seceng remains untested, but the hwrng half works well. If you are available soon, I would appreciate some assistance in testing this please.

Thanks!

**Comment by alexhorner** [comment]
Sorry, forgot to rebase so that the merge will succeed. All sorted!

File: arch/riscv/boot/dts/bouffalolab/bl808.dtsi [hunk 10]
```diff
@@ -67,6 +67,37 @@
...
+			gpio-ranges = <&pinctrl 0 0 46>;
+			bflb,npins = <46>;
+
+			status = "disabled";
+
+			interrupt-controller;
+			#interrupt-cells = <2>;
+			interrupts-extended = <&ipclic BFLB_IPC_SOURCE_M0
+					BFLB_IPC_DEVICE_GPIO IRQ_TYPE_EDGE_RISING>;
+			mboxes = <&ipclic BFLB_IPC_SOURCE_M0 BFLB_IPC_DEVICE_GPIO>;
```
Comment (arm000) [inline arch/riscv/boot/dts/bouffalolab/bl808.dtsi]: You can remove this line, this driver doesn't use mailboxes

File: arch/riscv/boot/dts/bouffalolab/bl808-pine64-ox64.dts [hunk 11]
```diff
@@ -9,7 +9,7 @@
 
 / {
 	model = "Pine64 Ox64";
-	compatible = "sipeed,m1s", "bflb,bl808";
+	compatible = "pine64,ox64", "bflb,bl808";
```
Comment (arm000) [inline arch/riscv/boot/dts/bouffalolab/bl808-pine64-ox64.dts:12]: Some of these changes are unrelated to the pinctrl driver, it will likely make integration with buildroot_bouffalo if you refactor them into a separate patch

**Review by arm000 — CHANGES_REQUESTED** [review]

**Comment by alexhorner** [comment]
Should be all resolved now
